elm-format-lib/src/Text/Parsec/Char.hs

```diff
@@ -54,5 +63,120 @@ anyChar             = satisfy (const True)
 satisfy :: (Char -> Bool) -> Parser Char
 satisfy = undefined
 
+
 string :: String -> Parser String
-string = undefined
+string = fmap decode . word8s . concatMap encodeChar
+
+
+word8s :: [Word8] -> Parser [Word8]
+word8s [] = return []
+word8s (w:ws) =
+  Parser $ EP.Parser $ \s@(EP.State _ pos end _ row col) cok _ cerr eerr ->
+    let
+      sourcePos = newPos "TODO - SourceName" row col
+
+      errEof _ _ = setErrorMessage (Expect (show (w:ws)))
+                    (newErrorMessage (SysUnExpect "") sourcePos)
+
+      errExpect x _ _ = setErrorMessage (Expect (show (w:ws)))
+                          (newErrorMessage (SysUnExpect (show x)) sourcePos)
+
+      walk [] s' = cok (w:ws) s'
+      walk (w':ws') s'@(EP.State _ pos' _ _ row' col') =
+        if pos' == end then
+          cerr row col errEof
+        else if EP.unsafeIndex pos' == w' then
+          walk ws' (updatePos w' s')
+        else
+          cerr row col (errExpect w')
+    in
+    if pos == end then
+      eerr row col errEof
+    else if EP.unsafeIndex pos == w then
+        walk ws (updatePos w s)
+    else
+      eerr row col (errExpect w)
+
+
+updatePos :: Word8 -> EP.State -> EP.State
+updatePos w (EP.State src pos end indent row col) =
+  let
+    (row', col') =
+      case w of
+        0x0a {- "\n" -} -> (row + 1, 1)
+
+        -- The doccumentation for `Text.Parsec.Char.updatePosChar` claims that
+        -- carrige return ("\r") increments row by 1, just like newline ("\n").
+        -- This doesn't appear to be the case from looking at the code:
+        -- https://hackage.haskell.org/package/parsec-3.1.14.0/docs/src/Text.Parsec.Pos.html#updatePosChar
+        --
+        -- Let's not devle into this unless it turns out that elm-format
+        -- needs it.
+        0x0d {- "\r" -} -> error "Can't handle carrige return"
```

Comment:
If I remember correctly, elm-format will never output CRs, but it does read them.  For ones in literal strings, it will convert them to an escape sequence.  And for ones elsewhere, it will treat them either as newlines or whitespace (I forget which).